### PR TITLE
etcdmain: fix the check in fallback-to-proxy case

### LINKED ISSUE #3018: advertise-client-urls has to be set if listen-client-urls is set when fallback to proxy
The behavior today:

```
$ etcd --listen-client-urls http://10.0.0.1:2379 -discovery xxx (-discovery-fallback proxy)
```

will fail anyway, regardless of whether discovery service is full:

```
2015/06/19 14:08:03 etcdmain: error verifying flags, -advertise-client-urls is required when -listen-client-urls is set explicitly. See 'etcd --help'.
2015/06/19 14:08:03 etcdmain: When listening on specific address(es), this etcd process must advertise accessible url(s) to each connected client.
```

This command will not fail at 2.0.10

@xiang90 points out that etcd should not fail due to unset advertise-client-urls if the discovery service is full and etcd falls back to proxy.

/cc @crawford This may be a regression, please check.

### PR DESCRIPTION
**Pull request by yichengq**
advertise-client-urls has to be set if listen-client-urls is set when
fallbacking to proxy, which breaks the behavior. Loosen the check to fix
it.

fixes #3018

### CODE REVIEW

File: etcdmain/config.go [hunk 1]
```diff
@@ -271,7 +271,9 @@ func (cfg *config) Parse(arguments []string) error {
 	}
 
 	// when etcd runs in member mode user needs to set -advertise-client-urls if -listen-client-urls is set.
-	if cfg.proxy.String() != proxyFlagOn {
+	// TODO: check this for joining through discovery service case
```
Comment (xiang90) [inline etcdmain/config.go]: TODO(yichengq):

**Comment by xiang90** [comment]
I think this deservers a test case.

**Comment by xiang90** [comment]
LGTM

**Comment by yichengq** [comment]
test added.

File: etcdmain/config_test.go [hunk 2]
```diff
@@ -212,6 +212,32 @@ func TestConfigParsingConflictClusteringFlags(t *testing.T) {
 	}
 }
 
+func TestConfigParsingMissedAdvertiseClientURLsFlag(t *testing.T) {
```
Comment (xiang90) [inline etcdmain/config_test.go]: can we also add the nil error case?

**Comment by yichengq** [comment]
added nil test. PTAL
It does find some missing part. It should not restrict when `proxy=readonly`.

File: etcdmain/config.go [hunk 3]
```diff
@@ -271,7 +271,10 @@ func (cfg *config) Parse(arguments []string) error {
 	}
 
 	// when etcd runs in member mode user needs to set -advertise-client-urls if -listen-client-urls is set.
-	if cfg.proxy.String() != proxyFlagOn {
+	// TODO(yichengq): check this for joining through discovery service case
+	mayFallbackToProxy := flags.IsSet(cfg.FlagSet, "discovery") && cfg.fallback.String() == fallbackFlagProxy
+	mayBeProxy := cfg.proxy.String() == proxyFlagOn || cfg.proxy.String() == proxyFlagReadonly || mayFallbackToProxy
```
Comment (xiang90) [inline etcdmain/config.go]: can we use proxy != off? that seems to be simpler and safer. (we might introduce more proxy flags)

**Comment by yichengq** [comment]
updated.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**